### PR TITLE
Bug 1997173: Add CRDs to the GVR incompatibility list

### PR DESCRIPTION
**Pull request by sseago**
Add CRDs to the GVR incompatibility list. Rather than warning
on the incompatibility of the resource apiversion itself (as we do
for non-CRDs), we warn for CRDs in use in the namespaces *if*
1) the CRD apiextensions apiversion is incompatible between clusters, and
2) the CRD in use doesn't exist in the destination cluster

### CODE REVIEW

**Comment by sseago** [comment]
A couple points:
1) I haven't yet been able to test this with a 4.9 dest cluster
2) I'm seeing that the CRD "compare resources" call isn't working right. It's reporting on CRDs that actually exist on dest. I need to investigate that and revise the PR.

**Comment by sseago** [comment]
Figured out the issue with 2). Bitten by the absence of "version" in v1 CRDs. I need to look at the "versions" slice instead. Will update the PR.

**Comment by sseago** [comment]
Updated. It's now ready for review. I've refactored the CRD list comparison to fix the version problems, as well as removed the dependency on v1beta1 availability in the source cluster for existing gvr checks. Also, I've been able to test this with a 4.9 target cluster.

File: pkg/gvk/gvk.go [hunk 1]
```diff
@@ -300,14 +485,19 @@ func (r *Compare) compareResources(src []*metav1.APIResourceList, dst []*metav1.
...
 	for _, srcList := range src {
 		missing := []metav1.APIResource{}
+		matchingDstResourceList := findResourceList(srcList.GroupVersion, dst)
 		for _, resource := range srcList.APIResources {
 			if cohabitator, found := r.CohabitatingResources[resource.Name]; found {
-				if cohabitator.Seen {
-					continue
+				gv, err := schema.ParseGroupVersion(srcList.GroupVersion)
+				if err != nil &&
+					(gv.Group == cohabitator.groupResource1.Group || gv.Group != cohabitator.groupResource2.Group) {
```
Comment (alaypatel07) [inline pkg/gvk/gvk.go:493]: little confused why this is required. @sseago can you elaborate on what is happening here?

Comment (sseago) [reply pkg/gvk/gvk.go:493]: @alaypatel07 The previous code identified cohabitating resources by the two groups they cohabitated in, but the code here that identified cohabitators and passed on them didn't bother to check the group. If I were to create a CRD that I call "deployments" in my own "sseago.example.com" group, it would be flagged as cohabitating with "deployments.extensions" and "deployments.apps", which seems like a bug. This change checks the two groupresource fields (`groupResource1` and `groupResource2` ) to make sure that we only flag cohabitating resources that actually belong to the cohabitating groups. There was actually no code in mig-controller that read the values of these two fields until I added this check.

Comment (sseago) [reply pkg/gvk/gvk.go:493]: I need to check the unit tests -- looks like this change may have broken a test.

Comment (alaypatel07) [reply pkg/gvk/gvk.go:493]: ack, I think you might want

```
				if err == nil &&
					(gv.Group == cohabitator.groupResource1.Group || gv.Group == cohabitator.groupResource2.Group) 
```

Comment (alaypatel07) [reply pkg/gvk/gvk.go:493]: also, @sseago if you are fixing this bug here, likely means we need the same bug fix in velero as well.

https://github.com/vmware-tanzu/velero/blob/bd5ba66e42e48ca0338e01d7c2d4375eea68e052/pkg/backup/item_collector.go#L211

Comment (sseago) [reply pkg/gvk/gvk.go:493]: @alaypatel07 Good catch on the backwards err nil check. Yes, if we succeed in parsing GV, then we use GV to compare with cohabitator Group. And yeah, I'll create an issue and PR upstream in velero. I'd forgotten that this code was taken from Velero. They're also not using the cohabitator groupResources for anything (other than logging).

Comment (sseago) [reply pkg/gvk/gvk.go:461]: Ugh. I shouild have said good catch on *both* backwards checks. Turns out the failing unit test actually did the right thing. Fixing both of these problems made the existing unit test work succeed again.

**Comment by sseago** [comment]
PR updated again. Fix from review applied. Also, I realized this morning that I didn't update the PR with the more significant refactoring work from yesterday, so that's now added to the commit as well.

File: pkg/gvk/gvk_test.go [hunk 2]
```diff
@@ -149,7 +149,7 @@ func Test_compareResources(t *testing.T) {
 							{
 								Name:       "deployments",
 								Namespaced: true,
-								Group:      "apps",
+								Group:      "extensions",
```
Comment (djwhatle) [inline pkg/gvk/gvk_test.go:152]: @sseago why did this test need to change? Was it not doing what it claimed to do previously?

Comment (sseago) [reply pkg/gvk/gvk_test.go:152]: @djwhatle So I first changed it when trying to debug an issue. But looking at it again (after realizing it wasn't really needed there), it looks like an error. `APIVersion` is "extensions/v1beta1". Group and Version of items in the `APIResources` slice should match this. I think it was a copy-and-paste oversight in the test data generation which doesn't really affect test results.

**Review by djwhatle — APPROVED** [review]
VISACK. I haven't been able to spin up a 4.9 cluster successfully to test with. @alaypatel07 I think you remember the specifics of these code paths better than I do after the deep dive you did a while back, would like to have you take one more look before we merge.

File: pkg/gvk/gvk.go [hunk 3]
```diff
@@ -153,6 +223,59 @@ func collectPreferredResources(discovery discovery.DiscoveryInterface) ([]*metav
 	return resources, nil
 }
 
+// collectPreferredResources collects all preferred namespaced scoped apiResources from the cluster
+func preferredCRDResource(discovery discovery.DiscoveryInterface) ([]*metav1.APIResourceList, error) {
```
Comment (alaypatel07) [inline pkg/gvk/gvk.go:227]: NIT: rename this to `collectPreferredCRDResources`? I think the doc comment needs an update as well

Comment (sseago) [reply pkg/gvk/gvk.go:227]: @alaypatel07 Makes sense -- although the singular "resource" was deliberate, since we will be returning exactly one resource in the slice -- the preferred GV for "customresourcedefinitions" -- should I call it `collectPreferredCRDResource` to indicate that, or should I make it plural like all the rest since it returns a slice, even though it will have a length of one?

**Review by alaypatel07 — CHANGES_REQUESTED** [review]
I wonder if adding a test case for the specific corner case will be helpful if we ever have to revisit this. i.e a test case where there are three `deployments` each belonging to different groups, 2 of which is apps and extension and third is user defined.

File: pkg/gvk/gvk.go (hunk 3 above)
Comment (alaypatel07) [reply pkg/gvk/gvk.go:227]: `collectPreferredCRDResource ` works

File: pkg/gvk/gvk.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/gvk/gvk.go]: +1, I should have pointed out both the errors. I was a little confused as to how best to communicate that so ended up pasting the exact string that worked

**Comment by sseago** [comment]
@alaypatel07 It might make sense to add an entry to the compareResources  test, actually. I"m currently trying to figure out how to add a test for my velero fix upstream, but that one is more challenging since I have to somehow create an actual deployment that's not in apps/extensions. For the compareResources test, we're not attempting to use any backup logic, so it should be easy enough to add it there. I'll do that this afternoon.

**Comment by sseago** [comment]
@alaypatel07 unit test case for the cohabitating fix added. I confirmed that the new case fails if I revert the cohabitating fix, and it succeeds with it.

**Review by alaypatel07 — APPROVED** [review]
VISIACK
